acme/challenge.go

```diff
@@ -573,32 +602,39 @@ func (dc *dns01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo validat
 	}
 	h := sha256.Sum256([]byte(expectedKeyAuth))
 	expected := base64.RawURLEncoding.EncodeToString(h[:])
-	var found bool
+
+	if len(txtRecords) == 0 {
+		e := errors.Errorf("no TXT record found at '%s'", record)
+		up.Error = DNSErr(e).ToACME()
+		return up, nil
+	}
+
 	for _, r := range txtRecords {
 		if r == expected {
-			found = true
-			break
-		}
-	}
-	if !found {
-		if err = dc.storeError(db,
-			RejectedIdentifierErr(errors.Errorf("keyAuthorization "+
-				"does not match; expected %s, but got %s", expectedKeyAuth, txtRecords))); err != nil {
-			return nil, err
+			up.Validated = time.Now().UTC()
+			up.Status = StatusValid
+			up.Error = nil
+			up.Retry = nil
+			return up, nil
 		}
-		return dc, nil
 	}
 
-	// Update and store the challenge.
-	upd := &dns01Challenge{dc.baseChallenge.clone()}
-	upd.Status = StatusValid
-	upd.Error = nil
-	upd.Validated = time.Now().UTC()
+	up.Status = StatusInvalid
+	e := errors.Errorf("keyAuthorization does not match; expected %s, but got %s",
+		expectedKeyAuth, txtRecords)
+	up.Error = IncorrectResponseErr(e).ToACME()
+	return up, nil
+}
 
-	if err := upd.save(db, dc); err != nil {
-		return nil, err
+// KeyAuthorization creates the ACME key authorization value from a token
```

Comment:
This method is only used in this file. Probably doesn't need to be exported.

Reply:
I didn't add this, just moved it down to the bottom. I can make it private if it's not intended to be called by others.